### PR TITLE
expression, plan: move typeinfer of agg funcs to expression rewriter

### PR DESCRIPTION
**Pull request by XuHuaiyu**
we try to move typeinfer process from expression/typeinferer.go to plan/expression_rewriter.go
avoid traversing ast tree twice.

PTAL @hanfei1991

### CODE REVIEW

File: expression/typeinferer.go [hunk 1]
```diff
@@ -52,8 +52,6 @@ func (v *typeInferrer) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
 
 func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
 	switch x := in.(type) {
-	case *ast.AggregateFuncExpr:
-		v.aggregateFunc(x)
```
Comment (hanfei1991) [inline expression/typeinferer.go]: The defination of this function should also be removed.

File: plan/typeinfer_test.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package plan_test
```
Comment (hanfei1991) [inline plan/typeinfer_test.go]: Miss license.

File: expression/aggregation.go [hunk 3]
```diff
@@ -557,8 +555,7 @@ func (af *avgFunction) Clone() AggregationFunction {
 // GetType implements AggregationFunction interface.
 func (af *avgFunction) GetType() *types.FieldType {
 	ft := types.NewFieldType(mysql.TypeNewDecimal)
-	ft.Charset = charset.CharsetBin
-	ft.Collate = charset.CollationBin
+	types.SetBinChsClnFlag(ft)
```
Comment (zz-jason) [inline expression/aggregation.go]: forget to set `ft.Flen` ?

**Comment by zz-jason** [comment]
LGTM

@hanfei1991 PTAL

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Pull request merged**